### ISSUE TITLE
nccl AllToAll hangs unexpectedly

### ISSUE THREAD

**Issue by gavin1332** [NVIDIA/nccl#613]
I have a multithreading code to do AllToAll, and it bring me confused information in my settings:
1. V100 node with nccl_2.9.6-1+cuda11.0_x86_64
2. Only the first 4 GPUs are used with full-mesh nvlinks
2. 4 threads in a single process with one device per thread
3. the stream and nccl_comm is correctly created after cudaSetDevice

The AllToAll code is as follows:
```
  template <typename T>
  void AllToAll(const T* send, T* recv, size_t send_cnt_each, cudaStream_t stream) {
    NCCL_CHECK(ncclGroupStart());
    for (size_t i = 0; i < nranks_; ++i) {
      NCCL_CHECK(ncclRecv(recv, send_cnt_each, NCCLDataType<T>::type, i, nccl_comm_, stream));
      NCCL_CHECK(ncclSend(send, send_cnt_each, NCCLDataType<T>::type, i, nccl_comm_, stream));
      send += send_cnt_each;
      recv += send_cnt_each;
    }  
    NCCL_CHECK(ncclGroupEnd());
  }
```
The information I have got:
1. If the 3rd thread enter ncclGroupStart in first and it always leaves ncclGroupEnd before other threads initializing their communications, which cause others hanging at ncclGroupEnd with 0% GPU-Utility, and the 3rd thread also hangs with 100% GPU-Utility when it synchronizes its stream .
2. If other threads run ahead of the 3rd one (occasionally or let the 3rd thread sleep for a second), everything works well and the result of alltoall is correct. 
3. We could run ncclAllReduce successfully with the `send` buffer right before ncclGroupStart
4. It also works well if we use only 2 threads 

The NCCL INFO is as follows:
```
tbds-9-45-101-175:68718:68755 [0] NCCL INFO Bootstrap : Using eth1:9.45.101.175<0>
tbds-9-45-101-175:68718:68755 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
tbds-9-45-101-175:68718:68755 [0] NCCL INFO NET/IB : No device found.
tbds-9-45-101-175:68718:68755 [0] NCCL INFO NET/Socket : Using [0]eth1:9.45.101.175<0> [1]br-c2aff00eaa0a:172.18.0.1<0>
tbds-9-45-101-175:68718:68755 [0] NCCL INFO Using network Socket
NCCL version 2.9.6+cuda11.0
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Trees [0] 3/-1/-1->2->0 [1] 0/-1/-1->2->3 [2] 3/-1/-1->2->0 [3] 0/-1/-1->2->3 [4] 3/-1/-1->2->0 [5] 0/-1/-1->2->3 [6] 3/-1/-1->2->0 [7] 0/-1/-1->2->3
tbds-9-45-101-175:68718:68753 [1] NCCL INFO Trees [0] -1/-1/-1->1->3 [1] 3/-1/-1->1->-1 [2] -1/-1/-1->1->3 [3] 3/-1/-1->1->-1 [4] -1/-1/-1->1->3 [5] 3/-1/-1->1->-1 [6] -1/-1/-1->1->3 [7] 3/-1/-1->1->-1
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Channel 00/08 :    0   2   3   1
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Setting affinity for GPU 2 to ff,ffff0000,00ffffff
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Channel 01/08 :    0   2   1   3
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Trees [0] 1/-1/-1->3->2 [1] 2/-1/-1->3->1 [2] 1/-1/-1->3->2 [3] 2/-1/-1->3->1 [4] 1/-1/-1->3->2 [5] 2/-1/-1->3->1 [6] 1/-1/-1->3->2 [7] 2/-1/-1->3->1
tbds-9-45-101-175:68718:68753 [1] NCCL INFO Setting affinity for GPU 1 to ff,ffff0000,00ffffff
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Setting affinity for GPU 3 to ff,ffff0000,00ffffff
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Channel 02/08 :    0   1   3   2
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Channel 03/08 :    0   3   1   2
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Channel 04/08 :    0   2   3   1
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Channel 05/08 :    0   2   1   3
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Channel 06/08 :    0   1   3   2
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Channel 07/08 :    0   3   1   2
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Trees [0] 2/-1/-1->0->-1 [1] -1/-1/-1->0->2 [2] 2/-1/-1->0->-1 [3] -1/-1/-1->0->2 [4] 2/-1/-1->0->-1 [5] -1/-1/-1->0->2 [6] 2/-1/-1->0->-1 [7] -1/-1/-1->0->2
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Setting affinity for GPU 0 to ff,ffff0000,00ffffff
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Channel 01 : 3[3e000] -> 0[1a000] via P2P/direct pointer
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Channel 00 : 2[3d000] -> 3[3e000] via P2P/direct pointer
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Channel 05 : 3[3e000] -> 0[1a000] via P2P/direct pointer
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Channel 04 : 2[3d000] -> 3[3e000] via P2P/direct pointer
tbds-9-45-101-175:68718:68753 [1] NCCL INFO Channel 03 : 1[1b000] -> 2[3d000] via P2P/direct pointer
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Channel 02 : 0[1a000] -> 1[1b000] via P2P/direct pointer
tbds-9-45-101-175:68718:68753 [1] NCCL INFO Channel 07 : 1[1b000] -> 2[3d000] via P2P/direct pointer
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Channel 06 : 0[1a000] -> 1[1b000] via P2P/direct pointer
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Channel 02 : 2[3d000] -> 0[1a000] via P2P/direct pointer
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Channel 00 : 3[3e000] -> 1[1b000] via P2P/direct pointer
tbds-9-45-101-175:68718:68753 [1] NCCL INFO Channel 01 : 1[1b000] -> 3[3e000] via P2P/direct pointer
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Channel 00 : 0[1a000] -> 2[3d000] via P2P/direct pointer
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Channel 03 : 2[3d000] -> 0[1a000] via P2P/direct pointer
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Channel 03 : 3[3e000] -> 1[1b000] via P2P/direct pointer
tbds-9-45-101-175:68718:68753 [1] NCCL INFO Channel 02 : 1[1b000] -> 3[3e000] via P2P/direct pointer
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Channel 01 : 0[1a000] -> 2[3d000] via P2P/direct pointer
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Channel 06 : 2[3d000] -> 0[1a000] via P2P/direct pointer
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Channel 04 : 3[3e000] -> 1[1b000] via P2P/direct pointer
tbds-9-45-101-175:68718:68753 [1] NCCL INFO Channel 05 : 1[1b000] -> 3[3e000] via P2P/direct pointer
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Channel 04 : 0[1a000] -> 2[3d000] via P2P/direct pointer
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Channel 07 : 2[3d000] -> 0[1a000] via P2P/direct pointer
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Channel 07 : 3[3e000] -> 1[1b000] via P2P/direct pointer
tbds-9-45-101-175:68718:68753 [1] NCCL INFO Channel 06 : 1[1b000] -> 3[3e000] via P2P/direct pointer
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Channel 05 : 0[1a000] -> 2[3d000] via P2P/direct pointer
tbds-9-45-101-175:68718:68753 [1] NCCL INFO Channel 00 : 1[1b000] -> 0[1a000] via P2P/direct pointer
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Channel 02 : 3[3e000] -> 2[3d000] via P2P/direct pointer
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Channel 01 : 2[3d000] -> 1[1b000] via P2P/direct pointer
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Channel 03 : 0[1a000] -> 3[3e000] via P2P/direct pointer
tbds-9-45-101-175:68718:68753 [1] NCCL INFO Channel 04 : 1[1b000] -> 0[1a000] via P2P/direct pointer
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Channel 06 : 3[3e000] -> 2[3d000] via P2P/direct pointer
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Channel 05 : 2[3d000] -> 1[1b000] via P2P/direct pointer
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Channel 07 : 0[1a000] -> 3[3e000] via P2P/direct pointer
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Connected all rings
tbds-9-45-101-175:68718:68753 [1] NCCL INFO Connected all rings
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Connected all rings
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Connected all rings
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Channel 01 : 2[3d000] -> 3[3e000] via P2P/direct pointer
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Channel 02 : 2[3d000] -> 3[3e000] via P2P/direct pointer
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Channel 03 : 2[3d000] -> 3[3e000] via P2P/direct pointer
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Channel 05 : 2[3d000] -> 3[3e000] via P2P/direct pointer
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Channel 06 : 2[3d000] -> 3[3e000] via P2P/direct pointer
tbds-9-45-101-175:68718:68753 [1] NCCL INFO Channel 00 : 1[1b000] -> 3[3e000] via P2P/direct pointer
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Channel 02 : 0[1a000] -> 2[3d000] via P2P/direct pointer
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Channel 07 : 2[3d000] -> 3[3e000] via P2P/direct pointer
tbds-9-45-101-175:68718:68753 [1] NCCL INFO Channel 03 : 1[1b000] -> 3[3e000] via P2P/direct pointer
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Channel 03 : 0[1a000] -> 2[3d000] via P2P/direct pointer
tbds-9-45-101-175:68718:68753 [1] NCCL INFO Channel 04 : 1[1b000] -> 3[3e000] via P2P/direct pointer
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Channel 06 : 0[1a000] -> 2[3d000] via P2P/direct pointer
tbds-9-45-101-175:68718:68753 [1] NCCL INFO Channel 07 : 1[1b000] -> 3[3e000] via P2P/direct pointer
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Channel 07 : 0[1a000] -> 2[3d000] via P2P/direct pointer
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Channel 00 : 2[3d000] -> 0[1a000] via P2P/direct pointer
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Channel 01 : 3[3e000] -> 1[1b000] via P2P/direct pointer
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Channel 01 : 2[3d000] -> 0[1a000] via P2P/direct pointer
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Channel 02 : 3[3e000] -> 1[1b000] via P2P/direct pointer
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Channel 04 : 2[3d000] -> 0[1a000] via P2P/direct pointer
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Channel 05 : 3[3e000] -> 1[1b000] via P2P/direct pointer
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Channel 05 : 2[3d000] -> 0[1a000] via P2P/direct pointer
tbds-9-45-101-175:68718:68752 [0] NCCL INFO Connected all trees
tbds-9-45-101-175:68718:68752 [0] NCCL INFO threadThresholds 8/8/64 | 32/8/64 | 8/8/512
tbds-9-45-101-175:68718:68752 [0] NCCL INFO 8 coll channels, 8 p2p channels, 2 p2p channels per peer
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Channel 06 : 3[3e000] -> 1[1b000] via P2P/direct pointer
tbds-9-45-101-175:68718:68752 [0] NCCL INFO comm 0x7fabf0000fc0 rank 0 nranks 4 cudaDev 0 busId 1a000 - Init COMPLETE
tbds-9-45-101-175:68718:68753 [1] NCCL INFO Connected all trees
tbds-9-45-101-175:68718:68753 [1] NCCL INFO threadThresholds 8/8/64 | 32/8/64 | 8/8/512
tbds-9-45-101-175:68718:68753 [1] NCCL INFO 8 coll channels, 8 p2p channels, 2 p2p channels per peer
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Channel 00 : 3[3e000] -> 2[3d000] via P2P/direct pointer
tbds-9-45-101-175:68718:68753 [1] NCCL INFO comm 0x7fabcc22ee50 rank 1 nranks 4 cudaDev 1 busId 1b000 - Init COMPLETE
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Channel 01 : 3[3e000] -> 2[3d000] via P2P/direct pointer
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Channel 03 : 3[3e000] -> 2[3d000] via P2P/direct pointer
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Channel 04 : 3[3e000] -> 2[3d000] via P2P/direct pointer
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Channel 05 : 3[3e000] -> 2[3d000] via P2P/direct pointer
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Channel 07 : 3[3e000] -> 2[3d000] via P2P/direct pointer
tbds-9-45-101-175:68718:68754 [2] NCCL INFO Connected all trees
tbds-9-45-101-175:68718:68756 [3] NCCL INFO Connected all trees
tbds-9-45-101-175:68718:68754 [2] NCCL INFO threadThresholds 8/8/64 | 32/8/64 | 8/8/512
tbds-9-45-101-175:68718:68756 [3] NCCL INFO threadThresholds 8/8/64 | 32/8/64 | 8/8/512
tbds-9-45-101-175:68718:68754 [2] NCCL INFO 8 coll channels, 8 p2p channels, 2 p2p channels per peer
tbds-9-45-101-175:68718:68756 [3] NCCL INFO 8 coll channels, 8 p2p channels, 2 p2p channels per peer
tbds-9-45-101-175:68718:68754 [2] NCCL INFO comm 0x7fabc422ee50 rank 2 nranks 4 cudaDev 2 busId 3d000 - Init COMPLETE
tbds-9-45-101-175:68718:68756 [3] NCCL INFO comm 0x7fabd022ee50 rank 3 nranks 4 cudaDev 3 busId 3e000 - Init COMPLETE
```

Any ideas?

**Comment by sjeaugey** [comment]
Does `NCCL_LAUNCH_MODE=GROUP` solve the hang?

**Comment by gavin1332** [comment]
It works! But what's the difference?

**Comment by sjeaugey** [comment]
Well, it's a complicated matter.

If you run gdb on your stuck process you should see that one thread has already launched its NCCL kernel, then there is a chain of dependencies from other CUDA calls your application, made which is blocked because of the blocked NCCL kernel (waiting for other NCCL kernels to start), and which prevents other threads from launching their own NCCL kernel.

Typically, something like:
```
cudaCall1();
ncclAllReduce();
cudaCall2();
````
can cause such hangs, if cudaCall2 on one GPU blocks cudaCall1 on another GPU.

`NCCL_LAUNCH_MODE=GROUP` waits until all threads within the process have called NCCL to start all kernels for all GPUs. The launch time is much higher because one thread is now responsible for launching on all GPUs, which needs contexts switches, but on the other hand there is much less risk for applications to experience hangs.

**Comment by gavin1332** [comment]
Thanks for your detailed explanation!

**Comment by gavin1332** [comment]
As NCCL_LAUNCH_MODE is declared to be deprecated in the newest version of NCCL, is there any suggestion for our future work to get rid of this flag?

**Comment by sjeaugey** [comment]
We intended to remove the group mode since CUDA cooperative Group multi-device launch is deprecated, but it looks like it won't go away. Even if we remove the CGMD launch, we will still keep the simple loop launch (hence the `NCCL_LAUNCH_MODE=GROUP`) to help workaround issues like the one you have. So you should not worry about that going away until there is no way for applications to cause hangs.

Now, to solve the problem, you could also make sure there is no inter-GPU cuda dependency by e.g. adding a barrier before the NCCL call.

**Comment by gavin1332** [comment]
> by e.g. adding a barrier before the NCCL call.

Actually, I have tested in experimentally to insert a ncclAllReduce before the `ncclGroupStart` as a barrier, but it does not work.

**Comment by gavin1332** [comment]
Got! Thanks for your help

**Comment by xw285cornell** [comment]
> The launch time is much higher because one thread is now responsible for launching on all GPUs, which needs contexts switches, but on the other hand there is much less risk for applications to experience hangs.

@sjeaugey just curious, why cudaCall2 on one device will block the cudaCall1 on the other device?

**Comment by sjeaugey** [comment]
> why cudaCall2 on one device will block the cudaCall1 on the other device?
There are a variety of reasons why it could happen, but in general, it is tied to memory management and pointer sharing when CUDA peer access is enabled (P2P). When you enable peer access between two GPUs within the same process, it means that all pointers valid on one GPU are also valid on the other GPU. Hence, each time you allocate or free memory on one GPU, it has to be also valid on the other GPUs which have peer access enabled to that GPU. To ensure that, all memory changes on one GPU (cudaMalloc, cudaFree) may have to wait for all kernels to complete on all GPUs, change the memory mappings, then unblock all GPUs.

**Comment by xw285cornell** [comment]
@sjeaugey thanks! This makes sense. Another qq, does this group launch mode still hurts performance, besides the barriers across the threads? Is it still using the cooperative API that can be quite a bit of overhead?

**Comment by sjeaugey** [comment]
We no longer use the cooperative group launch and use a loop of normal launches instead, which means we're faster but we could also be less atomic at launch, risking that a CUDA operation happening on an independent side thread gets inserted between two launches and causes a hang.

**Comment by yinghai** [comment]
> We no longer use the cooperative group launch and use a loop of normal launches instead

@sjeaugey in which NCCL version do we begin to avoid cooperative launch?

**Comment by chrisxcai** [comment]
Hi @sjeaugey, we currently have a hanging issue which might be caused by NCCL in a multi-threading multi-rank training environment, here is our setting:

We break down the training iteration into a few phases or stages, each phase is run by a separate thread on a GPU, and the output of a phase will be the input to the next phase, in order to pipeline these phases. Say we have 4 phases and 8 GPUs, we will have a total of 4 * 8  = 32 threads. Each of the phase will perform some CUDA/NCCL operations, for example, phase 1 might be doing some data copying from CPU to GPU, phase 2 issues NCCL AlltoAll, phase 3 issues NCCL allReduce, etc.

We are using NCCL 2.17.1, and already set NCCL_LAUNCH_MODE="group" after reading up this thread. Currently we found when we use larger number of GPUs (> 8), the training threads often hang on some CUDA/NCCL operations. For 8 GPUs we experienced much less often hanging. 

For example, attached is a sample GDB thread backtrace dump on 8 GPUs on the same host out of a 128-GPU run, if we take a look at what operation each of these training threads are hanging on, we will find:

Dataloading threads(e.g. Thread 89) hanging on cudaMemcpyAsync ()
```
#11 0x00007f8e45265215 in cudaMemcpyAsync () from target:/usr/local/fbcode/platform010/lib/libcudart.so.11.4.108
#12 0x0000000037d17cd7 in at::native::copy_kernel_cuda (iter=..., non_blocking=<optimized out>) at ./fbcode/caffe2/c10/cuda/CUDAStream.h:89
```
7 out of the 8 threads issuing NCCL AlltoAll(e.g. Thread 159) hanging on groupLaunch:

```
Thread 159 (Thread 0x7f8511600640 (LWP 2418485)):
#0  0x00007f8e6f310127 in sched_yield () at ../sysdeps/unix/syscall-template.S:120
#1  0x000000003a00fb83 in groupLaunch (job_=0x7f85115fae18) at fbsource/comm.h:364
#2  0x000000003a00f154 in ncclGroupEndInternal () at third-party/nccl/v2.17.1-1/src/group.cc:422
#3  0x000000003a00ece4 in ncclGroupEnd () at third-party/nccl/v2.17.1-1/src/group.cc:106
#4  0x0000000039f128fa in c10d::ProcessGroupNCCL::groupEnd() ()
#5  0x0000000039ee91b4 in c10d::ProcessGroupNCCL::endCoalescing() ()
```
1 out of the 8 threads issuing NCCL AlltoAll(e.g. Thread 159) hanging on other CUDA operation like cudaMemcpyAsync():
```
Thread 150 (Thread 0x7f852fe00640 (LWP 2415510)):
#0  0x00007f8e4749268d in ?? () from target:/usr/local/fbcode/platform010/lib/libcuda.so.1
#1  0x00007f8e473714d6 in ?? () from target:/usr/local/fbcode/platform010/lib/libcuda.so.1
#2  0x00007f8e47590fec in ?? () from target:/usr/local/fbcode/platform010/lib/libcuda.so.1
#3  0x00007f8e475918b1 in ?? () from target:/usr/local/fbcode/platform010/lib/libcuda.so.1
#4  0x00007f8e476b8b50 in ?? () from target:/usr/local/fbcode/platform010/lib/libcuda.so.1
#5  0x00007f8e47331973 in ?? () from target:/usr/local/fbcode/platform010/lib/libcuda.so.1
#6  0x00007f8e47331e81 in ?? () from target:/usr/local/fbcode/platform010/lib/libcuda.so.1
#7  0x00007f8e47332de8 in ?? () from target:/usr/local/fbcode/platform010/lib/libcuda.so.1
#8  0x00007f8e474ffd21 in ?? () from target:/usr/local/fbcode/platform010/lib/libcuda.so.1
#9  0x00007f8e45241359 in ?? () from target:/usr/local/fbcode/platform010/lib/libcudart.so.11.4.108
#10 0x00007f8e45216a0d in ?? () from target:/usr/local/fbcode/platform010/lib/libcudart.so.11.4.108
#11 0x00007f8e45265215 in cudaMemcpyAsync () from target:/usr/local/fbcode/platform010/lib/libcudart.so.11.4.108
```


Threads issuing NCCL AllReduce(e.g. Thread 160 ) also hanging on groupLaunch:
```
Thread 160 (Thread 0x7f8510c00640 (LWP 2418486)):
#0  0x00007f8e6f310127 in sched_yield () at ../sysdeps/unix/syscall-template.S:120
#1  0x000000003a00fb83 in groupLaunch (job_=0x7f8510bfae18) at fbsource/comm.h:364
#2  0x000000003a00f154 in ncclGroupEndInternal () at third-party/nccl/v2.17.1-1/src/group.cc:422
#3  0x000000003a00ece4 in ncclGroupEnd () at third-party/nccl/v2.17.1-1/src/group.cc:106
#4  0x0000000039f5f607 in torch::cuda::nccl::AutoNcclGroup::~AutoNcclGroup() ()
```

[gdb.txt](https://github.com/NVIDIA/nccl/files/12443928/gdb.txt)


Do you think the hanging is likely caused by the fact we are issuing concurrent NCCL operations from multiple threads on the same GPUs? Please let us know if anything needs to be clarified, thanks. cc @xw285cornell
